lib/streamlit/elements/chat.py

```diff
@@ -0,0 +1,120 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Optional, cast
+
+from typing_extensions import Final
+
+from streamlit import runtime
+from streamlit.elements.utils import check_callback_rules, check_session_state_rules
+from streamlit.errors import StreamlitAPIException
+from streamlit.proto.ChatInput_pb2 import ChatInput as ChatInputProto
+from streamlit.proto.Common_pb2 import StringTriggerValue as StringTriggerValueProto
+from streamlit.proto.RootContainer_pb2 import RootContainer
+from streamlit.runtime.metrics_util import gather_metrics
+from streamlit.runtime.scriptrunner import get_script_run_ctx
+from streamlit.runtime.state import (
+    WidgetArgs,
+    WidgetCallback,
+    WidgetKwargs,
+    register_widget,
+)
+from streamlit.type_util import Key, to_key
+
+if TYPE_CHECKING:
+    from streamlit.delta_generator import DeltaGenerator
+
+DISALLOWED_CONTAINERS_ERROR_TEXT = "`st.chat_input()` can't be used inside an `st.expander`, `st.form`, `st.tabs`, `st.columns`, `st.container`, or `st.sidebar`."
```

Comment:
I think it would actually be good to allow it within `st.container` as long as it only has `st.container` or `st.main` blocks as parents. Not doing that makes it a lot harder for developers to write any higher-level API for this. And I think there isn't any specific reason why we should prevent this. The version before worked fine with `st.container` since the `Vertical` block wasn't added yet, so it wasn't part of the parent block types.